applications/StructuralMechanicsApplication/custom_utilities/perturb_geometry_subgrid_utility.cpp

```diff
@@ -0,0 +1,148 @@
+// KRATOS  ___|  |                   |                   |
+//       \___ \  __|  __| |   |  __| __| |   |  __| _` | |
+//             | |   |    |   | (    |   |   | |   (   | |
+//       _____/ \__|_|   \__,_|\___|\__|\__,_|_|  \__,_|_| MECHANICS
+//
+//  License:		 BSD License
+//					 license: structural_mechanics_application/license.txt
+//
+//  Main authors:    Manuel Messmer
+//
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "custom_utilities/perturb_geometry_subgrid_utility.h"
+#include "custom_utilities/node_search_utility.h"
+#include "utilities/builtin_timer.h"
+#include "utilities/openmp_utils.h"
+#include "utilities/parallel_utilities.h"
+#include "utilities/variable_utils.h"
+
+namespace Kratos
+{
+
+int PerturbGeometrySubgridUtility::CreateRandomFieldVectors(){
+    KRATOS_TRY;
+
+    int num_of_nodes = mrInitialModelPart.NumberOfNodes();
+    // Search radius. Defines the minimum distance among the nodes in the subgrid
+    double radius = mMinDistanceSubgrid;
+
+    // Get all nodes
+    ModelPart::NodesContainerType nodes = mrInitialModelPart.Nodes();
+    // Construct and initialize searcher class
+    NodeSearchUtility searcher(nodes);
+
+    // Define the reduced space vector
+    std::vector<ModelPart::NodeType::Pointer> reduced_space_nodes;
+    ResultNodesContainerType  results;
+
+    BuiltinTimer reduced_space_timer;
+    // Mark all nodes as unvisited
+    VariableUtils().SetFlag(VISITED, false, mrInitialModelPart.Nodes());
+
+    // Generate reduced space
+    // It is assumed that the reduced space contains less the 50% of all ndoes
+    reduced_space_nodes.reserve((int)0.5*num_of_nodes);
+    const auto it_node_begin = mrInitialModelPart.NodesBegin();
+    for(int i = 0; i < num_of_nodes; ++i){
+        auto it_node = it_node_begin + i;
+        if(!it_node->Is(VISITED)) {
+            it_node->Set(VISITED,true);
+            reduced_space_nodes.push_back(&(*it_node));
+            results = {};
+            searcher.SearchNodesInRadius(*it_node, radius, results);
+            for(std::size_t j = 0; j < results.size(); ++j){
+                results[j]->Set(VISITED,true);
+            }
+        }
+    }
+    reduced_space_nodes.shrink_to_fit();
```

Comment:
```suggestion
```

I would remove this since it is a local variable

Otherwise you have two memory related operations (those are expensive)